Scene writer sets

donboie opened the pull request:
Write sets into caches. 

Optionally there is also an attempt to rationalize  RendererAlgo::parallelProcessLocations & SceneAlgo::parallelTraverse and use a parallel algorithm in the SceneWriter.

johnhaddon commented on include/GafferScene/SceneAlgo.inl:
```diff
@@ -66,43 +67,41 @@ class TraverseTask : public tbb::task
+				spawn( *t );
+			}
+			wait_for_all();
+
```
I should have made a todo comment for this, but I was thinking it would be really useful to call a method on the parent functor here, passing it the vector of child functors. This way it could gather results from children and percolate them back up to the root. For instance, this would allow matchingPaths to be implemented without any locks at all - each child could compute the PathMatcher relative to its location, and then the parent could combine them into a PathMatcher at its location and so on up the tree.

I guess we could introduce this later by using some template magic to call the method if it exists and not if it doesn't? Maybe we'd want that magic anyway, to keep things simple for the simple case.

donboie replied:
Great. I'll give this a go in another PR?

johnhaddon replied:
Cool, sounds good.

johnhaddon commented on src/GafferScene/SceneWriter.cpp:
```diff
@@ -53,85 +56,101 @@ using namespace GafferScene;
 {
 
-void writeLocation( const GafferScene::ScenePlug *scene, const ScenePlug::ScenePath &scenePath, Context *context, IECore::SceneInterface *output, double time )
+tbb::mutex g_writerMutex;
```
As far as I'm aware, it should be safe to write to different SceneCache files concurrently, so this would ideally be a per-file lock rather than a global one. Not the most important thing probably, since I don't think we do any parallel dispatch currently (well, we do it parallel across the farm, but not parallel per process).

johnhaddon commented on src/GafferScene/SceneWriter.cpp:
```diff
@@ -53,85 +56,101 @@ using namespace GafferScene;
-		ConstCompoundObjectPtr globals = scene->globalsPlug()->getValue();
-		output->writeAttribute( "gaffer:globals", globals.get(), time );
-	}
+		ContextPtr context = new Context( *Context::current() );
```
This context will be unnecessary once we're computing sets upfront or using `ScenePlug::set()`, but I should mention that this is a very expensive way of making a temporary context. `Context::EditableScope` is the preferred way - it's a good bit quicker and is also designed such that it can be optimised further in the future. The only reason to construct a new context like this is if you need its lifetime to be longer than the scope of the function - typically just in the UI.

johnhaddon commented on src/GafferScene/SceneWriter.cpp:
```diff
@@ -53,85 +56,101 @@ using namespace GafferScene;
-		if( pathMatcher.match( scenePath ) == Filter::ExactMatch )
+		tbb::mutex::scoped_lock l( g_writerMutex );
+
+		if ( !scenePath.empty() )
```
Since you once nitpicked my whitespacing, I'll take the same liberty :) The prevailing style in Gaffer is : 

```
if( !scenePath.empty() )
{
}
```

johnhaddon commented on src/GafferScene/SceneWriter.cpp:
```diff
@@ -189,6 +189,27 @@ void SceneWriter::writeLocation( const GafferScene::ScenePlug *scene, const Scen
+		context->set( ScenePlug::setNameContextName, setNames[i] );
+		const PathMatcher& pathMatcher = scene->setPlug()->getValue()->readable();
+
+		if( pathMatcher.match( scenePath ) == Filter::ExactMatch )
```
Shouldn't this be `| Filter::ExactMatch` rather than `== Filter::ExactMatch`? Otherwise it won't write out the set membership if there are ancestors or descendants in the set. If that's the case, could you update the test so that it's capable of catching this error and make the fix?

donboie commented on include/GafferScene/SceneAlgo.h:
```diff
@@ -78,8 +78,30 @@ void matchingPaths( const Gaffer::IntPlug *filterPlug, const ScenePlug *scene, P
 
-/// Calls a functor on all paths in the scene
-/// The functor must take ( const ScenePlug*, const ScenePlug::ScenePath& ), and can return false to prune traversal
+/// Invokes the Functor at every location in the scene,
```
I checked gaffer and thought this was safe (perhaps barring minor performance implications) ones but you're right about other usages at IE requiring a single functor to accumulate results.

So I'll keep parallelTraverse in place and just add parallelProcessLocations

johnhaddon replied:
Cool - sounds good.

donboie commented:
Regarding performance I got a little side tracked by thinking about continuous performance testing here at IE  and didn't get round to generating the numbers for this PR & will generate some data today.

donboie commented on src/GafferScene/SceneWriter.cpp:
```diff
@@ -189,6 +189,27 @@ void SceneWriter::writeLocation( const GafferScene::ScenePlug *scene, const Scen
+
+	for( size_t i = 0, numSets = setNames.size(); i < numSets; ++i )
+	{
+		context->set( ScenePlug::setNameContextName, setNames[i] );
```
Great! I was considering caching the sets but wasn't sure if this was wrong headed in terms of how contexts work and how gaffer caches. Thanks for the explanation.

donboie commented on src/GafferScene/SceneWriter.cpp:
```diff
@@ -189,6 +189,27 @@ void SceneWriter::writeLocation( const GafferScene::ScenePlug *scene, const Scen
+	for( size_t i = 0, numSets = setNames.size(); i < numSets; ++i )
+	{
+		context->set( ScenePlug::setNameContextName, setNames[i] );
+		const PathMatcher& pathMatcher = scene->setPlug()->getValue()->readable();
```
What a terrible mistake - thanks!

donboie commented:
My current tests indicate no change in performance as it's still executing on only one thread. I suspect this is because I'm not calculating the bounds in parallel and will investigate further before this PR should be looked at again.

johnhaddon commented:
Out of interest, what sort of scene are you testing with? Because we're parallel per-location and only visit child locations after processing parents, we can be single threaded at the roots of the hierarchy, so if there's a particularly expensive operation there you might expect to see what you're describing...

donboie commented:
Sorry for the tardy response but I've spent a little too long thinking "_I should be able to just figure this out by staring at stack traces_"

I'm testing with a scene which instances out planes and displaces each one using noise seeded by the instance id using OSL object. [test scene](https://gist.github.com/donboie/ddfcfe36422f656058303da3a00156a8)

I've tried a couple of things 

* switched from the IE instancer to the Gaffer instancer
  * made no difference
* removed writing the bounds entirely from the SceneWriter 
  * obviously generate invalid caches but had great core utilization &  fastest wall clock time.
  * Tried this as the  the gdb attach & break method of profiling seemed to imply I was calculating the bounds of the root location which would recurse to child bounds & OSL shading all in a single thread 
* write the bounds after all the children have been visited to try to improve on threading performance and the latest commit in this PR includes this change
  * as per you comment about adding function to call the parent functor when all child locations have been visited. 
  * This performed better but there seemed to be a significant serial processing step when after the child displaced meshes bounds / object had been computed. 

I'd hope to get to something like the SceneGadget update performance minus the serialization but it's no where near.

johnhaddon commented:
Thanks Don. Lot's to talk about here!

> I'm testing with a scene which instances out planes and displaces each one using noise seeded by the instance id using OSL object.

This is a classic worst-case for a deferred scene processing engine - the root bounds depending on the bounds of all the leaf locations. I suspect if you did the deformation upstream on the way into the instancer (using instancer:id to randomise, rather than scene:path) you would get much better performance, because the instancer has a special-case bounds implementation which threads internally. In any case, I don't think we should blame the SceneWriter for the problem - it's more fundamental than that.

> removed writing the bounds entirely from the SceneWriter
obviously generate invalid caches but had great core utilization & fastest wall clock time.

Did this really generate invalid caches? As far as I understood it, the SceneCache will automatically write bounds based on the bounds of the objects written to it, and automatically propagate the bounds up to the root. I think it even does special processing so that the bounds are expanded to account for any interpolation of rotations between samples.

I think the `writeBound()` call in the SceneWriter only exists because we give bounds to some things that Cortex doesn't know the bound of (lights for instance), and also because some nodes might deliberately manipulate the bound separately from the objects. It might be worth considering whether or not we could either never call `writeBound()`, or call it only for those special locations.

> write the bounds after all the children have been visited to try to improve on threading performance and the latest commit in this PR includes this change

This is a good idea, but I don't think it's something the SceneWriter should have to worry about. We actually have an [open ticket](#1149) about doing the bounds propagation in parallel in the nodes themselves, and I even have an implementation. This was confounded by it actually decreasing performance for some setups, something I was baffled by at the time. I now suspect it was due to the caching issue I'm working on right now, so I hope to revisit soon. Since then we've also introduced the PerformanceMonitor, so we're in a better position to diagnose any issues too.

> I'd hope to get to something like the SceneGadget update performance minus the serialization but it's no where near.

The SceneGadget only calculates bounds for non-expanded locations, so if you've expanded the scene it will be doing less work than the SceneWriter in any case.

Basically, I think what you've got here is good, and the parallelising of bounds propagation is a separate issue. Perhaps you could rejig your scene to put the deformation upstream of the instancer and verify that with this modified setup we do see a benefit? The only thing I'm really concerned about is whether or not the serial step bottlenecks to the point that we should be taking a more pipelined approach.

Assuming that shows some benefits I think we can go ahead and merge this, minus the final commit.
If you have time it might also be nice to bring back the commit where you updated SceneAlgo etc to use parallelProcessLocations behind the scenes, and deprecate the `parallelTraverse` methods in preference to the new one.

Cheers...
John

donboie commented:
Thanks for detailed notes.

> Did this really generate invalid caches? As far as I understood it, the SceneCache will automatically write bounds based on the bounds of the objects written to it, and automatically propagate the bounds up to the root. I think it even does special processing so that the bounds are expanded to account for any interpolation of rotations between samples.

 I just assumed the caches would be invalid if removed the writeBounds calls.

> If you have time it might also be nice to bring back the commit where you updated SceneAlgo etc to use parallelProcessLocations behind the scenes, and deprecate the parallelTraverse methods in preference to the new one.

I'm  not exactly sure what you're proposing here and I don't want to head off in the wrong direction. I can create a _filteredParallelProcessLocations_ function or overloaded _parallelProcessLocations_ with filter arguments which calls _parallelProcessLocations_ as part of this commit but are you saying I should update _filteredParallelTraverse_ to use _parallelProcessLocations_ ? There is one IE internal spot which calls _filteredParallelTraverse_ which will fail with the new copying functor. Once we have _parallelProcessLocations_ &  _filteredParallelProcessLocations_ in gaffer I can update the few internal usages to use the new functionality.

Also what is the convention for deprecating functions? A comment pointing the developer to the appropriate function?

johnhaddon commented:
> There is one IE internal spot which calls filteredParallelTraverse which will fail with the new copying functor.

Ah, yes, I'm getting ahead of myself aren't I, forgetting the copying problem I brought up before. You know what, let's just get this merged (still minus the final commit) and I'll look at the rest another time.

> Also what is the convention for deprecating functions? A comment pointing the developer to the appropriate function?

Yep, ideally using the \deprecated tag so it shows up in doxygen's deprecated page and can be easily grepped. But like I say, let's just get this merged and then I'll take a look at how we rearrange things later.

donboie commented:
running `time gaffer execute` on this updated [test scene](https://gist.github.com/donboie/d051b70c2ce6c8c8c62809ba5cc18519) displacing before the instancer gives:

```
 1 thread  : 3.918u 0.379s 0:05.01 85.4%	
 2 threads : 3.994u 0.359s 0:03.52 123.2%	
 4 threads : 4.352u 0.445s 0:03.23 148.2%	
 8 threads : 4.338u 0.435s 0:02.74 173.7%	
16 threads : 6.463u 0.552s 0:02.72 257.7%	
```

So hopefully this is good go now?

johnhaddon commented:
Great - thanks Don.

As a tip for the future, you can use `gaffer stats scene.gfr -task SceneWriter` to do timings with the stats app. This would give you timings for just the write task, without including the overhead of gaffer startup and script loading. You can then also use other flags like `-performanceMonitor` to get more details info about where the time is going...
